src/ai/backend/manager/services/permission_contoller/service.py

```diff
@@ -167,3 +171,12 @@ async def search_users_assigned_to_role(
             has_next_page=result.has_next_page,
             has_previous_page=result.has_previous_page,
         )
+
+    async def update_role_permissions(
+        self, action: UpdateRolePermissionsAction
+    ) -> UpdateRolePermissionsActionResult:
+        """Update role permissions using batch update."""
+        result = await self._repository._db_source.update_role_permissions(
```

Comment:
Instead of accessing the private member _db_source of self._repository directly, it would be better to add a dedicated method for this.